File: app/models/voter.py
"""Voter and VoterTag SQLAlchemy models."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from geoalchemy2 import Geometry, WKBElement
from sqlalchemy import ForeignKey, Index, SmallInteger, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import ARRAY, JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class Voter(Base):
    """Canonical voter record.

    Core fields are typed columns for efficient querying. Extra vendor-specific
    data (likelihood scores, ethnicity estimates, etc.) lives in the JSONB
    ``extra_data`` column.
    """

    __tablename__ = "voters"
    __table_args__ = (
        Index(
            "ix_voters_campaign_source",
            "campaign_id",
            "source_type",
            "source_id",
            unique=True,
        ),
        Index("ix_voters_campaign_party", "campaign_id", "party"),
        Index("ix_voters_campaign_precinct", "campaign_id", "precinct"),
        Index("ix_voters_campaign_reg_zip", "campaign_id", "registration_zip"),
        Index("ix_voters_campaign_last_name", "campaign_id", "last_name"),
        Index("ix_voters_campaign_mailing_zip", "campaign_id", "mailing_zip"),
        Index("ix_voters_campaign_mailing_city", "campaign_id", "mailing_city"),
        Index("ix_voters_campaign_mailing_state", "campaign_id", "mailing_state"),
    )

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    campaign_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("campaigns.id"), nullable=False
    )

    # Source tracking
    source_type: Mapped[str] = mapped_column(String(50), nullable=False)
    source_id: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Core fields
    first_name: Mapped[str | None] = mapped_column(String(255))
    middle_name: Mapped[str | None] = mapped_column(String(255))
    last_name: Mapped[str | None] = mapped_column(String(255))
    suffix: Mapped[str | None] = mapped_column(String(50))
    date_of_birth: Mapped[date | None] = mapped_column()
    gender: Mapped[str | None] = mapped_column(String(20))

    # Registration Address
    registration_line1: Mapped[str | None] = mapped_column(String(500))
    registration_line2: Mapped[str | None] = mapped_column(String(500))
    registration_city: Mapped[str | None] = mapped_column(String(255))
    registration_state: Mapped[str | None] = mapped_column(String(2))
    registration_zip: Mapped[str | None] = mapped_column(String(10))
    registration_zip4: Mapped[str | None] = mapped_column(String(4))
    registration_county: Mapped[str | None] = mapped_column(String(255))
    registration_apartment_type: Mapped[str | None] = mapped_column(String(20))

    # Mailing Address
    mailing_line1: Mapped[str | None] = mapped_column(String(500))
    mailing_line2: Mapped[str | None] = mapped_column(String(500))
    mailing_city: Mapped[str | None] = mapped_column(String(255))
    mailing_state: Mapped[str | None] = mapped_column(String(2))
    mailing_zip: Mapped[str | None] = mapped_column(String(10))
    mailing_zip4: Mapped[str | None] = mapped_column(String(4))
    mailing_country: Mapped[str | None] = mapped_column(String(100))
    mailing_type: Mapped[str | None] = mapped_column(String(20))

    # Political
    party: Mapped[str | None] = mapped_column(String(50))
    precinct: Mapped[str | None] = mapped_column(String(100))
    congressional_district: Mapped[str | None] = mapped_column(String(10))
    state_senate_district: Mapped[str | None] = mapped_column(String(10))
    state_house_district: Mapped[str | None] = mapped_column(String(10))
    registration_date: Mapped[date | None] = mapped_column()

    # Voting history (array of election identifiers)
    voting_history: Mapped[list | None] = mapped_column(ARRAY(String), default=list)

    # Propensity Scores
    propensity_general: Mapped[int | None] = mapped_column(SmallInteger)
    propensity_primary: Mapped[int | None] = mapped_column(SmallInteger)
    propensity_combined: Mapped[int | None] = mapped_column(SmallInteger)

    # Demographics
    ethnicity: Mapped[str | None] = mapped_column(String(100))
    age: Mapped[int | None] = mapped_column()
    spoken_language: Mapped[str | None] = mapped_column(String(100))
    marital_status: Mapped[str | None] = mapped_column(String(50))
    military_status: Mapped[str | None] = mapped_column(String(50))
    party_change_indicator: Mapped[str | None] = mapped_column(String(50))
    cell_phone_confidence: Mapped[int | None] = mapped_column(SmallInteger)

    # Geographic
    latitude: Mapped[float | None] = mapped_column()
    longitude: Mapped[float | None] = mapped_column()
    geom: Mapped[WKBElement | None] = mapped_column(
        Geometry(geometry_type="POINT", srid=4326, spatial_index=False),
        nullable=True,
    )
    # Household
    household_id: Mapped[str | None] = mapped_column(String(255))
    household_party_registration: Mapped[str | None] = mapped_column(String(50))
    household_size: Mapped[int | None] = mapped_column(SmallInteger)
    family_id: Mapped[str | None] = mapped_column(String(255))

    # Extras
    extra_data: Mapped[dict | None] = mapped_column(JSONB, default=dict)

    # Metadata
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now()
    )


class VoterTag(Base):
    """Campaign-scoped free-form voter tag."""

    __tablename__ = "voter_tags"
    __table_args__ = (
        UniqueConstraint("campaign_id", "name", name="uq_voter_tag_campaign_name"),
    )

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    campaign_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("campaigns.id"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)


class VoterTagMember(Base):
    """Join table linking voters to tags."""

    __tablename__ = "voter_tag_members"

    voter_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("voters.id"), primary_key=True
    )
    tag_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("voter_tags.id"), primary_key=True
    )

File: web/src/types/voter.ts
export interface Voter {
  id: string
  campaign_id: string
  source_type: string
  source_id: string | null

  first_name: string | null
  middle_name: string | null
  last_name: string | null
  suffix: string | null
  date_of_birth: string | null
  gender: string | null

  // Registration Address
  registration_line1: string | null
  registration_line2: string | null
  registration_city: string | null
  registration_state: string | null
  registration_zip: string | null
  registration_zip4: string | null
  registration_county: string | null
  registration_apartment_type: string | null

  // Mailing Address
  mailing_line1: string | null
  mailing_line2: string | null
  mailing_city: string | null
  mailing_state: string | null
  mailing_zip: string | null
  mailing_zip4: string | null
  mailing_country: string | null
  mailing_type: string | null

  // Political
  party: string | null
  precinct: string | null
  congressional_district: string | null
  state_senate_district: string | null
  state_house_district: string | null
  registration_date: string | null

  // Voting history
  voting_history: string[] | null

  // Propensity Scores
  propensity_general: number | null
  propensity_primary: number | null
  propensity_combined: number | null

  // Demographics
  ethnicity: string | null
  age: number | null
  spoken_language: string | null
  marital_status: string | null
  military_status: string | null
  party_change_indicator: string | null
  cell_phone_confidence: number | null

  // Geographic
  latitude: number | null
  longitude: number | null

  // Household
  household_id: string | null
  household_party_registration: string | null
  household_size: number | null
  family_id: string | null

  // Extras
  extra_data: Record<string, unknown> | null

  // Metadata
  created_at: string
  updated_at: string
}

export interface VoterFilter {
  search?: string
  party?: string
  parties?: string[]
  registration_city?: string
  registration_state?: string
  registration_zip?: string
  registration_county?: string
  precinct?: string
  congressional_district?: string
  age_min?: number
  age_max?: number
  gender?: string
  voted_in?: string[]
  not_voted_in?: string[]
  tags?: string[]
  tags_any?: string[]
  registered_after?: string
  registered_before?: string
  has_phone?: boolean

  // Propensity score ranges
  propensity_general_min?: number
  propensity_general_max?: number
  propensity_primary_min?: number
  propensity_primary_max?: number
  propensity_combined_min?: number
  propensity_combined_max?: number

  // Multi-select demographics
  ethnicities?: string[]
  spoken_languages?: string[]
  military_statuses?: string[]

  // Mailing address
  mailing_city?: string
  mailing_state?: string
  mailing_zip?: string

  logic?: "AND" | "OR"
}

export interface VoterCreate {
  first_name?: string
  middle_name?: string
  last_name?: string
  suffix?: string
  date_of_birth?: string
  gender?: string

  // Registration Address
  registration_line1?: string
  registration_line2?: string
  registration_city?: string
  registration_state?: string
  registration_zip?: string
  registration_zip4?: string
  registration_county?: string
  registration_apartment_type?: string

  // Mailing Address
  mailing_line1?: string
  mailing_line2?: string
  mailing_city?: string
  mailing_state?: string
  mailing_zip?: string
  mailing_zip4?: string
  mailing_country?: string
  mailing_type?: string

  // Political
  party?: string
  precinct?: string
  congressional_district?: string
  state_senate_district?: string
  state_house_district?: string
  registration_date?: string
  party_change_indicator?: string

  // Voting history
  voting_history?: string[]

  // Propensity Scores
  propensity_general?: number
  propensity_primary?: number
  propensity_combined?: number

  // Demographics
  ethnicity?: string
  age?: number
  spoken_language?: string
  marital_status?: string
  military_status?: string
  cell_phone_confidence?: number

  // Geographic
  latitude?: number
  longitude?: number

  // Household
  household_id?: string
  household_party_registration?: string
  household_size?: number
  family_id?: string

  // Extras
  extra_data?: Record<string, unknown>
  source_id?: string
  source_type?: string
}

export interface VoterUpdate extends Partial<VoterCreate> {}

export interface VoterSearchBody {
  filters: VoterFilter
  cursor?: string
  limit?: number
  sort_by?: string
  sort_dir?: "asc" | "desc"
}

export interface VoterInteraction {
  id: string
  campaign_id: string
  voter_id: string
  type: string
  payload: Record<string, unknown>
  created_by: string
  created_at: string
}
